File: C++/binarySearch.cpp
#include <iostream>
#include <vector>
using namespace std;

int binarySearch(vector<int> arr, int tar)          // Iterative code
{ 
    int st = 0, end = arr.size() - 1;

    while (st <= end)
    {
        // int mid = (st + end)/2;                 //Unoptimized for INT_MAX values
        int mid = st + (end - st) / 2;              // Optimized for INT_MAX values

        if (tar > arr[mid])
        {
            st = mid + 1;
        }
        else if (tar < arr[mid])
        {
            end = mid - 1;
        }
        else
        {
            return mid;
        }
    }

    return -1;
}

int main()
{
    vector<int> arr1 = {-1, 0, 3, 4, 6, 9, 12};
    int tar1 = 9;

    cout << binarySearch(arr1, tar1) << endl;

    vector<int> arr2 = {-1, 0, 3, 4, 6, 9, 12};
    int tar2 = 30;

    cout << binarySearch(arr2, tar2) << endl;
}

File: Java/binarySearch.java
public class binarySearch {

    public static int search(int arr[], int key) {
        int start = 0, end = arr.length - 1;

        while (start <= end) {
            int mid = (start + end) / 2;

            if (arr[mid] == key) {
                return mid;
            } else if (arr[mid] > key) {
                end = mid - 1;
            } else {
                start = mid + 1;
            }
        }
        return -1;
    }

    public static void main(String[] args) {

        int arr[] = { 1, 2, 4, 5, 7, 8, 10 };
        int key = 8;

        System.out.println("The Binary of arr is on index: " + search(arr, key));

    }
}